tests/apps/test_train.py

```diff
@@ -0,0 +1,213 @@
+
+        with monkeypatch.context() as m:
+            m.setattr(thewalrus.samples, "hafnian_sample_state", lambda *args, **kwargs: 1)
+            s_pnr = gbs._generate_samples(gbs.A_init_scaled, 10)
+
+        assert s_threshold == 0
+        assert s_pnr == 1
+
+    def test_add_A_init_samples_bad_shape(self, adj, n_mean, dim):
+        """Test that _add_A_init_samples raises a ValueError when input samples of incorrect
+        shape, i.e. of dim + 1 modes"""
+        gbs = train.VGBS(adj, n_mean, embedding, True)
+        s = np.ones((2, dim + 1))
+        with pytest.raises(ValueError, match="Must input samples of shape"):
+            gbs._add_A_init_samples(s)
+
+    def test_add_A_init_samples_none_there(self, adj, n_mean, dim):
+        """Test that _add_A_init_samples correctly adds samples"""
+        gbs = train.VGBS(adj, n_mean, embedding, True)
+        s = np.ones((2, dim))
+        gbs._add_A_init_samples(s)
+        assert np.allclose(gbs._A_init_samples, s)
+
+    def test_add_A_init_samples_already_there(self, adj, n_mean, dim):
+        """Test that _add_A_init_samples correctly adds more samples when some are already there"""
+        gbs = train.VGBS(adj, n_mean, embedding, True)
+        s = np.ones((2, dim))
+        gbs._A_init_samples = s
+        gbs._add_A_init_samples(s)
+        assert gbs._A_init_samples.shape == (4, dim)
+
+    def test_mean_photons_per_mode(self, n_mean, dim):
+        """Test that mean_photons_by_mode is correct when given a simple fully connected
+        adjacency matrix and an identity W. We expect each mode to have the same mean photon number
+        and for that to add up to n_mean."""
+        adj = np.ones((dim, dim))
+        params = np.zeros(dim)
+        gbs = train.VGBS(adj, n_mean, embedding, False)
+        n_mean_vec = np.round(gbs.mean_photons_by_mode(params), 5)
+
+        assert np.allclose(np.sum(n_mean_vec), n_mean)  # check that the vector sums to n_mean
+        assert len(np.unique(n_mean_vec)) == 1  # check that the vector is constant
+        assert np.allclose(n_mean_vec[0], n_mean / dim)  # check that elements have correct values
+
+    def test_mean_clicks_per_mode(self, n_mean, dim):
+        """Test that mean_clicks_by_mode is correct when given a simple fully connected
+        adjacency matrix and an identity W. We expect each mode to have the same mean click number
+        and for that to add up to n_mean."""
+        adj = np.ones((dim, dim))
+        params = np.zeros(dim)
+        gbs = train.VGBS(adj, n_mean, embedding, True)
+        n_mean_vec = np.round(gbs.mean_clicks_by_mode(params), 5)
+
+        assert np.allclose(np.sum(n_mean_vec), n_mean)  # check that the vector sums to n_mean
+        assert len(np.unique(n_mean_vec)) == 1  # check that the vector is constant
+        assert np.allclose(n_mean_vec[0], n_mean / dim)  # check that elements have correct values
+
+    def test_photons_clicks_comparison(self, n_mean, dim):
```

Comment:
This test seems a bit dangerous if we ever want to change `n_mean` or the number of modes. Maybe better to test that `n_mean_vec_click[0] < 1` and `n_mean_vec_photon[0] >= n_mean_vec_click[0]` ? Note the second inequality is true by construction

Reply:
Nice! I have changed the test to check that all elements of n_mean_vec_photon are larger than n_mean_vec_click and that all elements of n_mean_vec_click do not exceed one.